src/legacy/core_plugins/data/public/expressions/lib/loader.ts

```diff
@@ -0,0 +1,127 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Ast } from '@kbn/interpreter/common';
+import { Observable, Subject } from 'rxjs';
+import { first, share } from 'rxjs/operators';
+import { Adapters, InspectorSession } from '../../../../../../plugins/inspector/public';
+import { execute, ExpressionDataHandler } from './execute';
+import { ExpressionRenderHandler } from './render';
+import { RenderId, Data, IExpressionLoaderParams } from './_types';
+import { getInspector } from '../services';
+
+export class ExpressionLoader {
+  data$: Observable<Data>;
+  update$: Observable<any>;
+  render$: Observable<RenderId>;
+  events$: Observable<any>;
+
+  private dataHandler: ExpressionDataHandler;
+  private renderHandler: ExpressionRenderHandler;
+  private dataSubject: Subject<Data>;
+  private data: Data;
+
+  constructor(element: HTMLElement, expression: string | Ast, params: IExpressionLoaderParams) {
+    this.dataSubject = new Subject();
+    this.data$ = this.dataSubject.asObservable().pipe(share());
+
+    this.renderHandler = new ExpressionRenderHandler(element);
+    this.render$ = this.renderHandler.render$;
+    this.update$ = this.renderHandler.update$;
+    this.events$ = this.renderHandler.events$;
+
+    this.update$.subscribe(({ newExpression, newParams }) => {
+      this.update(newExpression, newParams);
+    });
+
+    this.data$.subscribe(data => {
+      this.render(data);
+    });
+
+    this.execute(expression, params);
+    // @ts-ignore
+    this.dataHandler = this.dataHandler;
+  }
+
+  destroy() {}
+
+  cancel() {
+    this.dataHandler.cancel();
+  }
+
+  getExpression(): string {
+    return this.dataHandler.getExpression();
+  }
+
+  getAst(): Ast {
+    return this.dataHandler.getAst();
+  }
+
+  getElement(): HTMLElement {
+    return this.renderHandler.getElement();
+  }
+
+  openInspector(title: string): InspectorSession {
+    return getInspector().open(this.inspect(), {
+      title,
+    });
+  }
+
+  inspect(): Adapters {
+    return this.dataHandler.inspect();
+  }
+
+  update(expression: string | Ast, params: IExpressionLoaderParams): Promise<RenderId> {
+    const promise = this.render$.pipe(first()).toPromise();
+
+    if (expression !== null) {
+      this.execute(expression, params);
+    } else {
+      this.render(this.data);
+    }
+    return promise;
+  }
+
+  private execute = async (
+    expression: string | Ast,
+    params: IExpressionLoaderParams
+  ): Promise<Data> => {
+    if (this.dataHandler) {
+      this.dataHandler.cancel();
+    }
+    this.dataHandler = execute(expression, params);
+    const data = await this.dataHandler.getData();
+    this.dataSubject.next(data);
+    return data;
+  };
+
+  private async render(data: Data): Promise<RenderId> {
```

Comment:
I might be missing something, but this is ultimately calling the renderer from the registry right (via the render handler?)

Are any renderers actually returning a `Promise<RenderId>` today? Otherwise we should perhaps be adding an `ExpressionRenderer` generic, similar to `ExpressionFunction`, so that this can actually be enforced.

Reply:
hmm ... here we call the `expression renderer` handler, which will return a promise which will always resolve with the renderId. There internally we call renderer from the registry and pass it the handlers. Once rendering is done the done() handler should be called (and is nowadays i think for all renderers) which then resolves the promise.